src/main/java/org/sagebionetworks/web/client/widget/ReactComponentV2.java

```diff
@@ -0,0 +1,280 @@
+    } else {
+      nonReactChildElements.forEach(Element::removeFromParent);
+      nonReactChildElements.clear();
+    }
+  }
+
+  private ReactElement<T, P> createReactElement() {
+    detachNonReactChildElements();
+    maybeUpdatePropsWithCallbackRef();
+    return React.createElement(
+      reactComponentType,
+      props,
+      getChildReactElements()
+    );
+  }
+
+  /**
+   * @return true iff there are children, and all children are React components
+   */
+  private boolean allChildrenAreReactComponents() {
+    boolean allChildrenAreReactComponents = getChildren().size() > 0;
+    for (Widget w : getChildren()) {
+      if (!(w instanceof ReactComponentV2)) {
+        allChildrenAreReactComponents = false;
+        break;
+      }
+    }
+    return allChildrenAreReactComponents;
+  }
+
+  private boolean isRenderedAsReactComponentChild() {
+    return (
+      getParent() instanceof ReactComponentV2 &&
+      ((ReactComponentV2<?, ?>) getParent()).allChildrenAreReactComponents()
+    );
+  }
+
+  /**
+   * This method returns the root ReactComponent widget, which is the only place where this React tree is attached to the DOM.
+   */
+  private ReactComponentV2<?, ?> getRootReactComponentWidget() {
+    if (isRenderedAsReactComponentChild()) {
+      return (
+        (ReactComponentV2<?, ?>) getParent()
+      ).getRootReactComponentWidget();
+    } else {
+      return this;
+    }
+  }
+
+  @Override
+  public HandlerRegistration addClickHandler(ClickHandler handler) {
+    return addDomHandler(handler, ClickEvent.getType());
+  }
+
+  private ReactElement<?, ?>[] getChildReactElements() {
+    if (this.allChildrenAreReactComponents()) {
+      // If all widget children are ReactNodes, get their ReactElements and add them as React children
+      List<ReactComponentV2<?, ?>> childWidgets = new ArrayList<>();
+      getChildren()
+        .forEach(w -> childWidgets.add(((ReactComponentV2<?, ?>) w)));
+
+      return childWidgets
+        .stream()
+        .map(ReactComponentV2::createReactElement)
+        .toArray(ReactElement<?, ?>[]::new);
+    } else {
+      return new ReactElement[0];
+    }
+  }
+
+  public void render() {
+    if (isRenderedAsReactComponentChild()) {
+      // This component will be rendered as a child of another React component, so destroy the root if one exists
+      destroyRoot();
+    }
+
+    // This component may be a React child of another component, so retrieve the root widget that renders this component tree.
+    ReactComponentV2<?, ?> componentToRender = getRootReactComponentWidget();
+
+    // Asynchronously schedule creating a root in case we queued up an unmount of the current root
+    // See https://stackoverflow.com/questions/73459382
+    Timer t = new Timer() {
+      @Override
+      public void run() {
+        componentToRender.createRoot();
+        // Create a fresh ReactElement tree and render it
+        componentToRender.root.render(componentToRender.createReactElement());
```

Comment:
When root.render is called, create a fresh tree of ReactElements. This is much more simple than the previous implementation (which did some crazy React.cloneElement + prop overrides). React will optimize changes when you pass a new tree to `render`